### PR TITLE
[19590] Change history remove change return statement

### PR DESCRIPTION
**Pull request by rsanchez15**
## Description

`History::remove_change` return statement was based on the fact that `remove_change_nts` return a `History::iterator` different than `end` iteration. However, `remove_change_nts` was also calling `erase` method and this will return the iterator pointing to the next element of the collection or `end` if that was the last element. Therefore, `remove_change_nts` could return `end` iterator and still be the expected result. 

This PR updates two main functions:

- `remove_change_nts`. This function will return the same iterator in case it could not remove the change.
- `History::remove_change`. This function will check if the change pointed by the iterator returned by `remove_change_nts` is the expected one (next iterator in the History colection) or not (same iterator).

This PR also expects to solve the following issue rising in Windows platforms testing any Fast DDS dependent software compiled in Debug mode.

```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\vector(192) : Assertion failed: vector iterators incompatible
```

This can be tested using any of the following tests in Fast DDS (only in Windows platforms and Debug cmake build type):

```
The following tests FAILED:
         88 - BlackboxTests_DDS_PIM.DDSBasic.MultithreadedPublisherCreation (Exit code 0xc0000409)
         89 - BlackboxTests_DDS_PIM.DDSBasic.MultithreadedReaderCreationDoesNotDeadlock (Exit code 0xc0000409)
         90 - BlackboxTests_DDS_PIM.DDSBasic.PidRelatedSampleIdentity (Exit code 0xc0000409)
         91 - BlackboxTests_DDS_PIM.DDSBasic.IgnoreParticipant (Exit code 0xc0000409)
         92 - BlackboxTests_DDS_PIM.DDSBasic.participant_ignore_local_endpoints (Exit code 0xc0000409)
         94 - BlackboxTests_DDS_PIM.DDSBasic.endpoint_custom_payload_pools (Exit code 0xc0000409)
```

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- *N/A* Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- *N/A* Any new/modified methods have been properly documented using Doxygen. 
- *N/A* Changes are ABI compatible. 
- *N/A* Changes are API compatible. 
- *N/A* New feature has been added to the `versions.md` file (if applicable).
- *N/A* New feature has been documented/Current behavior is correctly described in the documentation. 
    
- *N/A* Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by rsanchez15** [comment]
@richiprosima please test this

**Comment by rsanchez15** [comment]
@richiprosima please test this

File: src/cpp/rtps/history/WriterHistory.cpp [hunk 1]
```diff
@@ -185,7 +185,7 @@ History::iterator WriterHistory::remove_change_nts(
     {
         EPROSIMA_LOG_ERROR(RTPS_WRITER_HISTORY,
                 "You need to create a Writer with this History before removing any changes");
-        return changesEnd();
+        return remove_iterator_constness(removal);
     }
 
     if ( removal == changesEnd())
```
Comment (richiware) [inline src/cpp/rtps/history/WriterHistory.cpp:191]: The last `return` of this function should not be `changeEnd()`.

**Review by richiware — CHANGES_REQUESTED** [review]
`DataReaderHistory.cpp` implements `remove_change_nts` and uses it. Currently the function will never fail in the DataReader side, but it should be appreciated if you prepare these functions for this.

**Comment by richiware** [comment]
@richiprosima please test this

File: src/cpp/rtps/history/History.cpp [hunk 2]
```diff
@@ -246,6 +260,14 @@ bool History::get_earliest_change(
     return true;
 }
 
+History::iterator History::remove_iterator_constness(
+        const_iterator c_it)
+{
+    History::iterator it = changesBegin();
+    std::advance(it, std::distance<const_iterator>(it, c_it));
```
Comment (MiguelCompany) [inline src/cpp/rtps/history/History.cpp:267]:
```suggestion
    std::advance(it, std::distance<const_iterator>(m_changes.cbegin(), c_it));
```

File: test/mock/rtps/ReaderHistory/fastdds/rtps/history/ReaderHistory.h [hunk 3]
```diff
@@ -177,6 +182,22 @@ class ReaderHistory
...
+    {
+        return inner_change->sequenceNumber == outer_change->sequenceNumber &&
+               inner_change->writerGUID == outer_change->writerGUID;
+    }
+
+    iterator remove_iterator_constness(
+            const_iterator c_it)
+    {
+        iterator it = m_changes.begin();
+        std::advance(it, std::distance<const_iterator>(it, c_it));
```
Comment (MiguelCompany) [inline test/mock/rtps/ReaderHistory/fastdds/rtps/history/ReaderHistory.h:197]:
```suggestion
        std::advance(it, std::distance<const_iterator>(m_changes.cbegin(), c_it));
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Pull request merged**